proposed/0000-crust-refactor/crust-library.md

```diff
@@ -0,0 +1,250 @@
+    pub fn mapped_udp_socket<'c>(&'c self, bop_handle: &BopHandle)
+        -> BopResult<MappedUdpSocket>;
+    /// Connect a `MappedUdpSocket`.
+    pub fn utp_rendezvous_connect(&self, bop_handle: &BopHandle,
+                                         mapped_socket: MappedUdpSocket,
+                                         their_info: UdpRendezvousInfo)
+        -> BopResult<Stream>
+
+    /// Connecting
+    pub fn connect(&self, bop_handle: &BopHandle, endpoint: Endpoint)
+        -> BopResult<Stream>
+
+    /// Cacheing/Bootstrapping
+    pub fn cache_endpoint(&self, endpoint: Endpoint)
+    pub fn iter_endpoint_cache<'c>(&'c self) -> EndpointCacheIterator<'c>
+}
+```
+
+The service's beacon state (enabled or disabled) is controlled via a separate
+object. This is just to take advantage of the fact that - being a simple
+boolean - it's state can be represented at the type level. The user can enforce
+compile-time guarantees about the state of the beacon during particular parts
+of their code, or they can just stick the `ServiceBeaconState` in a `Mutex` and
+use it the same way they use the `ServiceController`.
+
+The full API (sans error handling) is given below.
+
+```rust
+impl Service {
+    /// Create a new service from the default configuration file.
+    pub fn new() -> Service;
+    /// Create a new service from the given config file.
+    pub fn new_from_config(cfg: Config) -> Service;
+}
+
+/// Rather than being a single, opaque type the `Service` is a struct with public fields, allowing
+/// it to be disassembled into components which can be used independently.
+struct Service {
+    /// Used to accept incoming connections.
+    pub acceptor: ServiceListener,
+    /// Used to control the state of the beacon.
+    pub beacon_state: ServiceBeaconState,
+    /// Used to receive endpoints broadcasted by peers on the local network.
+    pub beacon_receiver: ServiceBeaconReceiver,
+    /// Used to control the service.
+    pub controller: ServiceController,
+}
+
+impl ServiceListener {
+    pub fn accept(&mut self, bop_handle: &BopHandle) -> BopResult<Stream>;
+    pub fn incoming(&mut self, bop_handle: &BopHandle) -> Incoming;
+}
+
+impl Iterator for Incoming {
+    type Item = Stream;
+}
+
+impl ServiceBeaconState {
+    /// Test whether the beacon is enabled.
+    fn is_enabled(&self) -> bool,
+    /// Set the period between transmissions.
+    fn set_period(&mut self, period: Duration)
+    /// Enable to beacon.
+    fn enable(self) -> ServiceBeaconStateEnabled,
+    /// Disable to beacon.
+    fn disable(self) -> ServiceBeaconStateDisabled,
+}
+
+/// This is because some code may want a static guarantee that the beacon is held in a particular
+/// state. Usually, enforcing this sort of thing at the type-level would require typestate or
+/// dependent types. But as the state is only a boolean we can acheive this in rust by treating the
+/// states as two different types.
```

Comment:
To avoid duplicating the `struct ServiceBeaconState`, it might make sense to use a pattern like in [Hyper's `Response`](http://hyper.rs/hyper/hyper/server/response/struct.Response.html), with a generic parameter to distinguish between the states:
```rust
impl ServiceBeaconState<Dynamic> {
    fn is_enabled(&self) -> bool
    fn set_period(&mut self, period: Duration)
    fn enable(self) -> ServiceBeaconState<Enabled>
    fn disable(self) -> ServiceBeaconState<Disabled>
}

impl ServiceBeaconState<Enabled> {
    fn set_period(&mut self, period: Duration)
    fn disable(self) -> ServiceBeaconState<Disabled>
}

impl From<ServiceBeaconState<Enabled>> for ServiceBeaconState<Dynamic>
```